### PR TITLE
Created PluginCallbackTask.php

### PR DESCRIPTION
**Pull request by alejandroliu**
Since CallbackTask is deprecated, how about a PluginCallbackTask?  It is like CallbackTask in its simplicity but tracks its owner plugin so it is easier to clean-up.

I found CallbackTask convenient to use in my Plugins as it helped me avoid writing the same code in several places.

### CODE REVIEW

**Comment by shoghicp** [comment]
@alejandroliu I still have to implement it to be able to use lamdas

File: src/pocketmine/scheduler/PluginCallbackTask.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+	/**
+	 * @return callable
+	 */
+	public function getCallable(){
+		return $this->callable;
+	}
+
+	public function onRun($currentTicks){
+		call_user_func_array($this->callable, $this->args);
```
Comment (PEMapModder) [inline src/pocketmine/scheduler/PluginCallbackTask.php]: According to my test, this is faster.

```php
$c = $this->callable;
$c(...$this->args);
```

Also, consider pushing current ticks into the end of the args? Like

```php
$c = $this->callable;
$args = $this->args;
$args[] = $currentTick;
$c(...$args);

Comment (shoghicp) [inline src/pocketmine/scheduler/PluginCallbackTask.php]: @PEMapModder Just use a normal PluginTask. I'm not so sure about doing this, as it involves callbacks which... might confuse stack traces.

**Comment by alejandroliu** [comment]
Hi,

Added the suggestions by PEMapModder.  Also, tested with some lambda's and these also work.  I can use "function ($args) use ($something) ... " as the callable and works and keeps the right variables.  I am not sure how I could test it on a different thread, but my guess is that using lambda's for a callaback on a different thread will not work.

I also tested creating a couple of stack traces to see how it would look like.  In my opinion stack-traces still work, but I am enclosing a couple of example for your review:

# Trace 1 - Closure

	2015-04-28 08:14:01 [CRITICAL] Could not execute task pocketmine\scheduler\PluginCallbackTask: Undefined variable: z
	2015-04-28 08:14:01 [NOTICE] UndefinedVariableException: "Undefined variable: z" (E_NOTICE) in "/home/alex/PocketMine-MP/plugins/pocketmine-plugins/GrabBag/src/aliuly/grabbag/CmdAfterAt" at line 79
	2015-04-28 08:14:01 [DEBUG] #0 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/scheduler/PluginCallbackTask(61): aliuly\grabbag\CmdAfterAt->aliuly\grabbag\{closure}(integer 1, pocketmine\scheduler\PluginCallbackTask object, integer 118)
	2015-04-28 08:14:01 [DEBUG] #1 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/scheduler/TaskHandler(150): pocketmine\scheduler\PluginCallbackTask->onRun(integer 118)
	2015-04-28 08:14:01 [DEBUG] #2 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/scheduler/ServerScheduler(247): pocketmine\scheduler\TaskHandler->run(integer 118)
	2015-04-28 08:14:01 [DEBUG] #3 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/Server(2392): pocketmine\scheduler\ServerScheduler->mainThreadHeartbeat(integer 118)
	2015-04-28 08:14:01 [DEBUG] #4 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/Server(2188): pocketmine\Server->tick()
	2015-04-28 08:14:01 [DEBUG] #5 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/Server(2070): pocketmine\Server->tickProcessor()
	2015-04-28 08:14:01 [DEBUG] #6 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/Server(1748): pocketmine\Server->start()
	2015-04-28 08:14:01 [DEBUG] #7 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/PocketMine(462): pocketmine\Server->__construct(pocketmine\CompatibleClassLoader object, pocketmine\utils\MainLogger object, string /home/alex/PocketMine-MP/PocketMine-MP_v1.5/, string /home/alex/PocketMine-MP/PocketMine-MP_v1.5/, string /home/alex/PocketMine-MP/PocketMine-MP_v1.5/plugins/)

# Trace 2 - Member function

	2015-04-28 08:05:20 [CRITICAL] Could not execute task pocketmine\scheduler\PluginCallbackTask: Undefined variable: x
	2015-04-28 08:05:20 [NOTICE] UndefinedVariableException: "Undefined variable: x" (E_NOTICE) in "/home/alex/PocketMine-MP/plugins/pocketmine-plugins/GrabBag/src/aliuly/grabbag/CmdAfterAt" at line 61
	2015-04-28 08:05:20 [DEBUG] #0 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/scheduler/PluginCallbackTask(61): aliuly\grabbag\CmdAfterAt->runCommand(string say one, pocketmine\scheduler\PluginCallbackTask object, integer 139)
	2015-04-28 08:05:20 [DEBUG] #1 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/scheduler/TaskHandler(150): pocketmine\scheduler\PluginCallbackTask->onRun(integer 139)
	2015-04-28 08:05:20 [DEBUG] #2 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/scheduler/ServerScheduler(247): pocketmine\scheduler\TaskHandler->run(integer 139)
	2015-04-28 08:05:20 [DEBUG] #3 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/Server(2392): pocketmine\scheduler\ServerScheduler->mainThreadHeartbeat(integer 139)
	2015-04-28 08:05:20 [DEBUG] #4 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/Server(2188): pocketmine\Server->tick()
	2015-04-28 08:05:20 [DEBUG] #5 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/Server(2070): pocketmine\Server->tickProcessor()
	2015-04-28 08:05:20 [DEBUG] #6 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/Server(1748): pocketmine\Server->start()
	2015-04-28 08:05:20 [DEBUG] #7 /home/alex/PocketMine-MP/src/PocketMine-MP_v1.5/src/pocketmine/PocketMine(462): pocketmine\Server->__construct(pocketmine\CompatibleClassLoader object, pocketmine\utils\MainLogger object, string /home/alex/PocketMine-MP/PocketMine-MP_v1.5/, string /home/alex/PocketMine-MP/PocketMine-MP_v1.5/, string /home/alex/PocketMine-MP/PocketMine-MP_v1.5/plugins/)
	2015-04-28 08:07:01 [INFO] Scheduled for Tue, 28 Apr 2015 08:07:02 +0100

File: src/pocketmine/scheduler/PluginCallbackTask.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	/**
+	 * @param Plugin   $owner
+	 * @param callable $callable
+	 * @param array    $args
+	 */
+	public function __construct(Plugin $owner, callable $callable, array $args = []){
+		parent::__construct($owner);
+		$this->callable = $callable;
+		$this->args = $args;
+		$this->args[] = $this;
```
Comment (sekjun9878) [inline src/pocketmine/scheduler/PluginCallbackTask.php]: May I ask why you need to append additional variables to the end? I feel that it makes things complicated for no real need.

**Comment by alejandroliu** [comment]
If you are talking about the:

    $this->args[] = $this

That is to be compatible with the current `CallbackTask` functionality.   Adding the ticks was suggested by @PEMapModder .

**Comment by sekjun9878** [comment]
To be honest I don't think the Task should be injected at all. That should be controlled by the main plugin. Plugins can use use() and $this if they want access to the original scope anyway.

**Comment by alejandroliu** [comment]
I tend to agree, I frequently use CallbackTask and never seen the need to access the CallbackTask object itself.  I was just trying to keep the semantics the same.

Should that be removed then, but keep the ticks counter?

**Comment by sekjun9878** [comment]
IMHO I would say make it as simple as possible and remove both ticks and the Task object. CallbackTask is there for simplicity. It's not there to be a full replacement of creating your own Task.

**Comment by sekjun9878** [comment]
Would you be able to run rebase on the commits so they are one? So the commit history doesn't get 6 new commits. Also, @shoghicp ready for your confirmation then merge

**Comment by shoghicp** [comment]
@sekjun9878 Things like adding ticks is just trying to not to follow proper OOP conventions and just have it on simple callbacks (that need to be public). There is also no confirmation of the object being part of the Plugin.
This was a helper class that was created to move from 1.3 core to 1.4, and finally it's done that way. Users had a warning in the header, it'll be removed once we push to API 2.0 (it'll remove all deprecated calls or events, which doesn't meat that plugins won't be compatible)

**Comment by alejandroliu** [comment]
I did a rebase, so it is just one commit now.

**Comment by sekjun9878** [comment]
@shoghicp if you actually look at the code the ticks have been removed and the owner is now quite clear since it extends PluginTask. This isn't the original CallbackTask anymore.

**Comment by shoghicp** [comment]
@sekjun9878 I know. Just my opinion (and I left more on the IRC)

**Comment by alejandroliu** [comment]
So, can it be merged?  I think I am not the only developer who would use this class.

**Comment by rj0495** [comment]
I think this would be useful.

**Comment by sekjun9878** [comment]
This is solved by anonymous classes in PHP 7.

**Pull request closed without merging**